reference/preconditioner/isai_kernels.cpp

```diff
@@ -0,0 +1,232 @@
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+#include "core/matrix/csr_builder.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace reference {
+/**
+ * @brief The Isai preconditioner namespace.
+ *
+ * @ingroup isai
+ */
+namespace isai {
+
+
+template <typename ValueType, typename IndexType, typename Callable>
+void generic_generate(std::shared_ptr<const DefaultExecutor> exec,
+                      const matrix::Csr<ValueType, IndexType> *mtx,
+                      matrix::Csr<ValueType, IndexType> *inverse_mtx,
+                      Callable trs_solve)
+{
+    /*
+    Consider: aiM := inverse_mtx; M := mtx
+    I := Identity matrix
+    e(i) := unit vector i (containing all zeros except for row i, which is one)
+    S := Sparsity pattern of the desired aiM
+    S(i) := Sparsity pattern of row i of aiM (Set of non-zero columns)
+    D(i) := M[S(i), S(i)]
+    aiM := approximate inverse of M
+
+    Target: Solving (aiM * M = I)_{S} (aiM * M = I for the sparsity pattern S)
+    aiM[i, :] * D(i) = e(i)^T
+    <=> D(i)^T * aiM[i, :]^T = e(i)   =^ Triangular system (Trs)
+    Solve Trs, fill in aiM row by row (coalesced access)
+    */
+    const auto num_rows = mtx->get_size()[0];
+    const auto m_row_ptrs = mtx->get_const_row_ptrs();
+    const auto m_cols = mtx->get_const_col_idxs();
+    const auto m_vals = mtx->get_const_values();
+    auto i_row_ptrs = inverse_mtx->get_row_ptrs();
+    auto i_cols = inverse_mtx->get_col_idxs();
+    auto i_vals = inverse_mtx->get_values();
+
+    // expect mtx and inverse_mtx to have the same number of elems
+    const auto num_elems = mtx->get_num_stored_elements();
+    // Copy sparsity pattern of original into the inverse of L
+    std::copy_n(m_row_ptrs, num_rows + 1, i_row_ptrs);
+    std::copy_n(m_cols, num_elems, i_cols);
+
+    gko::Array<ValueType> rhs_array{exec};  // RHS for local trisystem
+    // memory for dense trisystem in column major:
+    gko::Array<ValueType> trisystem_array{exec};
+
+    for (size_type row = 0; row < num_rows; ++row) {
+        const auto i_row_begin = i_row_ptrs[row];
+        const auto i_row_end = i_row_ptrs[row + 1];
+        const auto i_row_elems = i_row_end - i_row_begin;
+
+        trisystem_array.resize_and_reset(i_row_elems * i_row_elems);
+        auto trisystem = trisystem_array.get_data();
+        std::fill_n(trisystem, i_row_elems * i_row_elems, zero<ValueType>());
+
+        for (size_type i = 0; i < i_row_elems; ++i) {
+            const auto col = i_cols[i_row_begin + i];
+            const auto m_row_end = m_row_ptrs[col + 1];
+            auto m_row_ptr = m_row_ptrs[col];
+            auto i_row_ptr = i_row_begin;
+            // Values in `trisystem` will be stored in transposed order
+            // to prevent the need for an explicit transpose.
+            // Since `trisystem` stores in column major, this is equivalent
+            // to storing it in row-major and reading in column major.
```

Comment:
Is it better now? This comment changed to:
```
// Instead of transposing `trisystem` after filling it
// (which the algorithm dictates), we simply store the values
// transposed, so we skip the explicit transpose.
// Since `trisystem` is stored in column major, we can write to
// it as if it was stored in row major (therefore, skipping the
// explicit transpose afterwards)
```

Reply:
LGTM